File: binaryTree.js
class Node {
	constructor(data) {
		this.data = data;
		this.right = null;
		this.left = null;
	}
}

class BinaryTree {
	constructor() {
		this.root = null;
	}

	insert(data) {
		let current = this.root;
		const DIRECTION = ["left", "right"];

		if (!current && current === this.root) {
			this.root = new Node(data);
			return;
		}
		while (current) {
			let index = Math.floor(Math.random() * 2);

			if (!current.left) {
				current.left = new Node(data);
				return;
			}
			if (!current.right) {
				current.right = new Node(data);
				return;
			}

			if (DIRECTION[index] === "left") {
				current = current.left;
			} else {
				current = current.right;
			}
		}
	}

	replaceRootRight(data) {
		const temp = this.root;
		this.root = new Node(data);
		this.root.right = temp;
	}

	replaceRootLeft(data) {
		const temp = this.root;
		this.root = new Node(data);
		this.root.left = temp;
	}

	insertRight(data) {
		let current = this.root;

		while (current) {
			if (!current.right) {
				break;
			}

			current = current.right;
		}

		current.right = new Node(data);
	}

	insertLeft(data) {
		let current = this.root;

		while (current) {
			if (!current.left) {
				break;
			}
			current = current.left;
		}

		current.left = new Node(data);
	}

	// inorderTraversal() {
	// 	//log out data in tree in left node, root, right node sequence
	// 	let current = this.root;
	// 	let pre;

	// 	while (current) {
	// 		//while current is not null
	// 		if (!current.left) {
	// 			// check if current.left is empty
	// 			console.log(current.data); //if current.left is empty display its data
	// 			current = current.right; // assign current to the current.right since we are at the end of the available left nodes
	// 		} else {
	// 			// if current.left is not empty
	// 			pre = current.left; // assign current.left to a new variable
	// 			while (pre.right && pre.right !== current) {
	// 				//loop till pre.right is null or pre.right is the same as current break out of loop otherwise
	// 				pre = pre.right; //while the condition is true assign the right node of the new variable to the variable
	// 			}
	// 			if (!pre.right) {
	// 				//check if pre.right is empty
	// 				pre.right = current; // if it is empty assign current from line 88 to pre.right
	// 				current = current.left; // reassign current to the left node of current from line 88 to pre
	// 			} else {
	// 				//if pre.right is not empty
	// 				pre.right = null; // reassign that node as empty
	// 				console.log(current.data); // log out current from line 88 data
	// 				current = current.right; // reassign current to current.right
	// 			}
	// 		}
	// 	}
	// }

	inorderTraversal(root, stack = []) {
		if (root) {
			this.inorderTraversal(root.left, stack);
			stack.push(root.data);
			this.inorderTraversal(root.right, stack);
		}

		return stack;
	}

	preorderTraversal(root, stack = []) {
		if (root) {
			stack.push(root.data);
			this.preorderTraversal(root.left, stack);
			this.preorderTraversal(root.right, stack);
		}

		return stack;
	}

	postorderTraversal(root, stack = []) {
		if (root) {
			this.postorderTraversal(root.left, stack);
			this.postorderTraversal(root.right, stack);
			stack.push(root.data);
		}

		return stack;
	}

	search(key) {
		const stack = this.inorderTraversal(this.root);
		return stack.some((value) => value === key);
	}
}

let d = new BinaryTree();
d.insert(1);
d.insert(2);
d.insert(3);
d.insert(4);
d.insert(5);

console.log(d.search(5));

// d.inorderTraversal();
// d.preorderTraversal();

File: binary_tree.py
'''A python implementation of a binary tree data structure'''
import random

class Node:
    '''Represents a node in a tree'''
    def __init__(self, data) -> None:
        self.data = data
        self.right = None
        self.left = None

    def __repr__(self) -> str:
        return f'{self.data} L-{self.left} R-{self.right}'

    def type_of(self) -> int:
        '''Return type of data stored in node'''
        return type(self.data)


class BinaryTree:
    '''Represents a binary tree data structure'''
    def __init__(self) -> None:
        self.root = None

    def __repr__(self) -> str:
        return f'Root {self.root}'

    def insert(self,data):
        '''Inserts data into the closest available node'''
        current = self.root
        direction = ['left','right']

        if current is self.root and not current:
            self.root = Node(data)
            return

        while current:
            if not current.left:
                current.left = Node(data)
                return

            if not current.right:
                current.right = Node(data)
                return

            index = random.randint(1, 2) % 2

            if direction[index] == 'left':
                current = current.left
            else:
                current = current.right

    def replace_root_left(self, data):
        '''Insert a new node to the root and move old node to the left leaf '''
        # replaces the root with a new node and
        # pushes the old root to the left node or the right node

        temp = self.root
        self.root = Node(data)
        self.root.left = temp

    def replace_root_right(self, data):
        '''Insert a new node to the root and move old node to the right leaf '''
        # replaces the root with a new node and
        # pushes the old root to the right node or the right node

        temp = self.root
        self.root = Node(data)
        self.root.right = temp

    def insert_left(self, data):
        '''Insert Data to the leftmost available node'''
        current = self.root

        while current:
            if not current.left:
                break
            current = current.left

        new_node = Node(data)
        current.left = new_node

    def insert_right(self,data):
        '''Insert data to the rightmost available node'''
        current = self.root

        while current:
            if not current.right:
                break
            current = current.right

        new_node = Node(data)
        current.right = new_node

    # def inorder(self): # TODO - Go through code //inorder traversal without recurssion or stack//
    #     '''Returns a list of data stored in the tree when traversed inorder'''
    #     current = self.root

    #     while current:
    #         if not current.left:
    #             print(current.data)
    #             current = current.right
    #         else:
    #             pre = current.left
    #             while pre.right and pre.right is not current:
    #                 pre = pre.right
    #             if not pre.right:
    #                 pre.right = current
    #                 current = current.left
    #             else:
    #                 pre.right = None
    #                 print(current.data)
    #                 current = current.right

    def inorder(self, root, stack = []): #inorder with recurssion
        '''Inorder traversal of a tree'''
        if root:
            self.inorder(root.left, stack)
            stack.append(root.data)
            self.inorder(root.right, stack)
        return stack

    def preorder(self,root, stack = []):
        '''Preorder traversal of a tree'''
        if root:
            stack.append(root.data)
            self.preorder(root.left,stack)
            self.preorder(root.right, stack)
        return stack

    def postorder(self,root, stack = []):
        '''Postorder traversal of a tree'''
        if root:
            self.postorder(root.left, stack)
            self.postorder(root.right, stack)
            stack.append(root.data)
        return stack

    def search(self,key):
        '''Search for the existence of key in tree'''
        current = self.root

        for data in self.inorder(current):
            if key == data:
                return True

        return False


c = BinaryTree()
c.insert(1)
c.insert(2)
c.insert(3)
c.insert(4)
c.insert(5)
c.insert(6)
c.insert(7)
print(c.search(7))
# d = c.inorder(c.root)
# print(d)
